### PR TITLE
Swipeable banner

### PR DESCRIPTION
**Pull request by nitaliano**
Adds support for being able to swipe the `InstructionsView` and update the camera and arrow to our current preview step.

@danesfeder @Guardiola31337  I will be cleaning this up more today, and some general feedback on this would be great

### CODE REVIEW

File: libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/instruction/InstructionView.java [hunk 1]
```diff
@@ -154,6 +166,7 @@ protected void onDetachedFromWindow() {
     cancelDelayedTransition();
   }
 
+
```
Comment (Guardiola31337) [inline libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/instruction/InstructionView.java]: NIT - unnecessary new line

File: libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/instruction/InstructionStepResources.java [hunk 2]
```diff
@@ -118,7 +134,7 @@ private void thenStep(LegStep upcomingStep, LegStep followOnStep, double current
 
   private boolean isValidStepDuration(LegStep upcomingStep, double currentDurationRemaining) {
     return upcomingStep.duration() <= VALID_UPCOMING_DURATION
-      && currentDurationRemaining <= VALID_CURRENT_DURATION;
+        && currentDurationRemaining <= VALID_CURRENT_DURATION;
```
Comment (Guardiola31337) [inline libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/instruction/InstructionStepResources.java]: NIT - check formatting

File: libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/NavigationViewModel.java [hunk 3]
```diff
@@ -413,7 +426,7 @@ private void deactivateInstructionPlayer() {
   private void playVoiceAnnouncement(Milestone milestone) {
     if (milestone instanceof VoiceInstructionMilestone) {
       SpeechAnnouncement announcement = SpeechAnnouncement.builder()
-        .voiceInstructionMilestone((VoiceInstructionMilestone) milestone).build();
+          .voiceInstructionMilestone((VoiceInstructionMilestone) milestone).build();
```
Comment (Guardiola31337) [inline libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/NavigationViewModel.java]: NIT - check formatting

File: libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/instruction/InstructionStepResources.java [hunk 4]
```diff
@@ -137,6 +153,6 @@ private boolean checkForNoneIndications(List<IntersectionLanes> lanes) {
 
   private boolean hasIntersections(LegStep step) {
     return step.intersections() != null
-      && step.intersections().get(0) != null;
+        && step.intersections().get(0) != null;
```
Comment (Guardiola31337) [inline libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/instruction/InstructionStepResources.java]: NIT - check formatting

File: libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/instruction/InstructionView.java [hunk 5]
```diff
@@ -435,9 +475,9 @@ private void bind() {
   private void initializeBackground() {
     if (android.os.Build.VERSION.SDK_INT < android.os.Build.VERSION_CODES.LOLLIPOP) {
       int navigationViewBannerBackgroundColor = ThemeSwitcher.retrieveThemeColor(getContext(),
-        R.attr.navigationViewBannerBackground);
+          R.attr.navigationViewBannerBackground);
```
Comment (Guardiola31337) [inline libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/instruction/InstructionView.java]: NIT - check formatting

File: libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/instruction/InstructionView.java [hunk 6]
```diff
@@ -435,9 +475,9 @@ private void bind() {
...
       int navigationViewListBackgroundColor = ThemeSwitcher.retrieveThemeColor(getContext(),
-        R.attr.navigationViewListBackground);
+          R.attr.navigationViewListBackground);
```
Comment (Guardiola31337) [inline libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/instruction/InstructionView.java]: NIT - check formatting

File: libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/instruction/InstructionView.java [hunk 7]
```diff
@@ -461,9 +501,9 @@ private void initializeBackground() {
    */
   private void initializeInstructionAutoSize() {
     TextViewCompat.setAutoSizeTextTypeUniformWithConfiguration(upcomingPrimaryText,
-      26, 30, 1, TypedValue.COMPLEX_UNIT_SP);
+        26, 30, 1, TypedValue.COMPLEX_UNIT_SP);
```
Comment (Guardiola31337) [inline libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/instruction/InstructionView.java]: NIT - check formatting

File: libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/instruction/InstructionView.java [hunk 8]
```diff
@@ -461,9 +501,9 @@ private void initializeBackground() {
...
     TextViewCompat.setAutoSizeTextTypeUniformWithConfiguration(upcomingSecondaryText,
-      20, 26, 1, TypedValue.COMPLEX_UNIT_SP);
+        20, 26, 1, TypedValue.COMPLEX_UNIT_SP);
```
Comment (Guardiola31337) [inline libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/instruction/InstructionView.java]: NIT - check formatting

File: libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/instruction/InstructionView.java [hunk 9]
```diff
@@ -474,7 +514,7 @@ private void initializeTurnLaneRecyclerView() {
     rvTurnLanes.setAdapter(turnLaneAdapter);
     rvTurnLanes.setHasFixedSize(true);
     rvTurnLanes.setLayoutManager(new LinearLayoutManager(getContext(),
-      LinearLayoutManager.HORIZONTAL, false));
+        LinearLayoutManager.HORIZONTAL, false));
```
Comment (Guardiola31337) [inline libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/instruction/InstructionView.java]: NIT - check formatting

File: libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/instruction/InstructionView.java [hunk 10]
```diff
@@ -539,67 +579,21 @@ private void showButtons() {
...
    * Looks to see if we have a new distance text.
    *
    * @param model provides distance text
    */
   private boolean newDistanceText(InstructionModel model) {
     return !upcomingDistanceText.getText().toString().isEmpty()
-      && !TextUtils.isEmpty(model.getStepResources().getStepDistanceRemaining())
-      && !upcomingDistanceText.getText().toString()
-      .contentEquals(model.getStepResources().getStepDistanceRemaining().toString());
+        && !TextUtils.isEmpty(model.getStepResources().getStepDistanceRemaining())
```
Comment (Guardiola31337) [inline libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/instruction/InstructionView.java]: NIT - check formatting

File: libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/instruction/InstructionView.java [hunk 11]
```diff
@@ -715,8 +718,8 @@ private void updateThenStep(InstructionModel model) {
    */
   private boolean shouldShowThenStep(InstructionModel model) {
     return turnLaneLayout.getVisibility() != VISIBLE
-      && model.getThenBannerText() != null
-      && model.getStepResources().shouldShowThenStep();
+        && model.getThenBannerText() != null
```
Comment (Guardiola31337) [inline libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/instruction/InstructionView.java]: NIT - check formatting

File: libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/map/NavigationMapboxMap.java [hunk 12]
```diff
@@ -312,6 +319,22 @@ public void resumeCamera(@NonNull Location location) {
     mapCamera.resume(location);
   }
 
+  public void moveCameraTo(LegStep step, List<Point> currentPoints, List<Point> upcomingPoints) {
```
Comment (Guardiola31337) [inline libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/map/NavigationMapboxMap.java]: Could you add Javadocs here?

File: libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/NavigationPresenter.java [hunk 13]
```diff
@@ -23,10 +26,15 @@ void onRecenterClick() {
     view.setSummaryBehaviorHideable(false);
     view.setSummaryBehaviorState(BottomSheetBehavior.STATE_EXPANDED);
     view.updateWaynameVisibility(true);
-    view.resetCameraPosition();
+    onResetCameraAndArrow();
```
Comment (Guardiola31337) [inline libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/NavigationPresenter.java]: Here when you say `Arrow` you mean the _puck_? If so, we should update. That's how we name it internally.

**Comment by danesfeder** [comment]
@nitaliano giving this some 👀 and testing right now.  Unfortunately, I'm running into the following crash when trying to swipe the banner - any ideas?  

```
java.lang.NullPointerException: Attempt to invoke virtual method 'com.mapbox.api.directions.v5.models.RouteLeg com.mapbox.services.android.navigation.v5.routeprogress.RouteProgress.currentLeg()' on a null object reference
        at com.mapbox.services.android.navigation.ui.v5.summary.list.InstructionListPresenter.isNewLeg(InstructionListPresenter.java:126)
        at com.mapbox.services.android.navigation.ui.v5.summary.list.InstructionListPresenter.addBannerInstructions(InstructionListPresenter.java:112)
        at com.mapbox.services.android.navigation.ui.v5.summary.list.InstructionListPresenter.updateBannerListWith(InstructionListPresenter.java:48)
        at com.mapbox.services.android.navigation.ui.v5.summary.list.InstructionListAdapter.updateBannerListWith(InstructionListAdapter.java:47)
        at com.mapbox.services.android.navigation.ui.v5.instruction.InstructionView.updateInstructionList(InstructionView.java:873)
        at com.mapbox.services.android.navigation.ui.v5.instruction.InstructionView.updateDataFromInstruction(InstructionView.java:787)
        at com.mapbox.services.android.navigation.ui.v5.instruction.InstructionView.showPreviewInstruction(InstructionView.java:922)
        at com.mapbox.services.android.navigation.ui.v5.instruction.InstructionView.handleSwipeLeft(InstructionView.java:888)
        at com.mapbox.services.android.navigation.ui.v5.instruction.InstructionView.access$1000(InstructionView.java:82)
        at com.mapbox.services.android.navigation.ui.v5.instruction.InstructionView$7.onSwipeLeft(InstructionView.java:935)
        at com.mapbox.services.android.navigation.ui.v5.instruction.InstructionSwipeGestureListener.onFling(InstructionSwipeGestureListener.java:26)
        at android.view.GestureDetector.onTouchEvent(GestureDetector.java:655)
        at com.mapbox.services.android.navigation.ui.v5.instruction.InstructionView$8.onTouch(InstructionView.java:949)
        at android.view.View.dispatchTouchEvent(View.java:11772)
        at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:2959)
        at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2636)
        at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:2965)
        at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2650)
        at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:2965)
        at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2650)
        at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:2965)
        at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2650)
        at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:2965)
        at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2650)
        at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:2965)
        at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2650)
        at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:2965)
        at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2650)
        at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:2965)
        at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2650)
        at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:2965)
        at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2650)
        at com.android.internal.policy.DecorView.superDispatchTouchEvent(DecorView.java:446)
        at com.android.internal.policy.PhoneWindow.superDispatchTouchEvent(PhoneWindow.java:1848)
        at android.app.Activity.dispatchTouchEvent(Activity.java:3315)
        at android.support.v7.view.WindowCallbackWrapper.dispatchTouchEvent(WindowCallbackWrapper.java:68)
        at com.android.internal.policy.DecorView.dispatchTouchEvent(DecorView.java:408)
        at android.view.View.dispatchPointerEvent(View.java:12015)
        at android.view.ViewRootImpl$ViewPostImeInputStage.processPointerEvent(ViewRootImpl.java:4866)
        at android.view.ViewRootImpl$ViewPostImeInputStage.onProcess(ViewRootImpl.java:4644)
```

**Comment by nitaliano** [comment]
@danesfeder I haven't dug into this much since my last rebase I'll check out it, and if you see anything else in the code that needs to change lmk.

**Comment by nitaliano** [comment]
Just looked into this I see what's going on, I will update the PR

**Comment by danesfeder** [comment]
Hey @nitaliano, we are going to go ahead and close this as the `InstructionView` is going to be refactored in #1746.  Thanks a lot for the work done to get this going.  We will definitely keep this code in mind moving forward.

**Pull request closed without merging**